File: ui/tests/integration/test_login_flow.mjs
// Test the complete staff login flow
import https from 'https';
import * as jose from 'jose';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

// Get env file from command line argument or default to .env.e2e
const envFileName = process.argv[2] || '.env.e2e';

// Load credentials from specified env file
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const envPath = path.resolve(__dirname, '../../', envFileName);
const envContent = fs.readFileSync(envPath, 'utf8');

function getEnvVar(name) {
    const match = envContent.match(new RegExp(`^${name}=(.*)$`, 'm'));
    if (!match) {
        throw new Error(`Environment variable ${name} not found in ${envFileName}`);
    }
    return match[1];
}

const FIREBASE_API_KEY = getEnvVar('VITE_FIREBASE_API_KEY');
const EMAIL = getEnvVar('E2E_TEST_EMAIL');
const PASSWORD = getEnvVar('E2E_TEST_PASSWORD');
const BACKEND_API_URL = getEnvVar('VITE_API_URL');

function makeRequest(url, options, postData = null) {
    return new Promise((resolve, reject) => {
        const req = https.request(url, options, (res) => {
            let data = '';
            res.on('data', (chunk) => data += chunk);
            res.on('end', () => {
                try {
                    const parsed = JSON.parse(data);
                    if (res.statusCode >= 200 && res.statusCode < 300) {
                        resolve(parsed);
                    } else {
                        reject(new Error(`HTTP ${res.statusCode}: ${JSON.stringify(parsed)}`));
                    }
                } catch (e) {
                    reject(new Error(`Parse error: ${data}`));
                }
            });
        });
        req.on('error', reject);
        if (postData) req.write(postData);
        req.end();
    });
}

async function firebaseLogin() {
    console.log('\n=== Step 1: Firebase Login ===');
    const url = `https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=${FIREBASE_API_KEY}`;
    
    const postData = JSON.stringify({
        email: EMAIL,
        password: PASSWORD,
        returnSecureToken: true
    });
    
    const options = {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'Content-Length': Buffer.byteLength(postData)
        }
    };
    
    const response = await makeRequest(url, options, postData);
    const idToken = response.idToken;
    console.log(`✓ Firebase ID Token obtained (length: ${idToken.length})`);
    return idToken;
}

async function getServiceAccount(firebaseToken) {
    console.log('\n=== Step 2: Get Service Account from Backend ===');
    const url = `${BACKEND_API_URL}/api/auth/service-account`;
    
    const options = {
        method: 'POST',
        headers: {
            'Authorization': `Bearer ${firebaseToken}`,
            'Content-Type': 'application/json'
        }
    };
    
    const response = await makeRequest(url, options);
    const serviceAccount = response.serviceAccount;
    console.log(`✓ Service Account obtained: ${serviceAccount.client_email}`);
    return serviceAccount;
}

async function getGoogleAccessToken(serviceAccount) {
    console.log('\n=== Step 3: Get Google OAuth Token ===');
    
    const alg = 'RS256';
    const privateKeyString = serviceAccount.private_key.includes('\\n') 
        ? serviceAccount.private_key.replace(/\\n/g, '\n') 
        : serviceAccount.private_key;
    
    const privateKey = await jose.importPKCS8(privateKeyString, alg);
    
    const jwt = await new jose.SignJWT({
        scope: 'https://www.googleapis.com/auth/spreadsheets https://www.googleapis.com/auth/drive.readonly'
    })
        .setProtectedHeader({ alg })
        .setIssuer(serviceAccount.client_email)
        .setSubject(serviceAccount.client_email)
        .setAudience('https://oauth2.googleapis.com/token')
        .setIssuedAt()
        .setExpirationTime('1h')
        .sign(privateKey);
    
    console.log('✓ JWT created');
    
    // Exchange JWT for access token
    const postData = `grant_type=urn:ietf:params:oauth:grant-type:jwt-bearer&assertion=${jwt}`;
    const options = {
        method: 'POST',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
            'Content-Length': Buffer.byteLength(postData)
        }
    };
    
    const response = await makeRequest('https://oauth2.googleapis.com/token', options, postData);
    const accessToken = response.access_token;
    console.log(`✓ Google Access Token obtained (length: ${accessToken.length})`);
    console.log(`  Expires in: ${response.expires_in} seconds`);
    return accessToken;
}

async function listSpreadsheets(accessToken) {
    console.log('\n=== Step 4: List Spreadsheets ===');
    
    const query = encodeURIComponent("mimeType='application/vnd.google-apps.spreadsheet' and trashed=false");
    const url = `https://www.googleapis.com/drive/v3/files?q=${query}&fields=files(id,name)`;
    
    const options = {
        method: 'GET',
        headers: {
            'Authorization': `Bearer ${accessToken}`,
            'Content-Type': 'application/json'
        }
    };
    
    const response = await makeRequest(url, options);
    const files = response.files || [];
    
    console.log(`\n✓ Found ${files.length} spreadsheets:`);
    files.forEach((file, idx) => {
        console.log(`  ${idx + 1}. ${file.name} (ID: ${file.id})`);
    });
    
    return files;
}

async function main() {
    console.log('============================================================');
    console.log('Testing Complete Staff Login Flow');
    console.log('============================================================');
    
    try {
        const firebaseToken = await firebaseLogin();
        const serviceAccount = await getServiceAccount(firebaseToken);
        const googleToken = await getGoogleAccessToken(serviceAccount);
        const spreadsheets = await listSpreadsheets(googleToken);
        
        console.log('\n============================================================');
        if (spreadsheets.length > 0) {
            console.log('✅ SUCCESS: Service account can access spreadsheets!');
            console.log(`   Total spreadsheets found: ${spreadsheets.length}`);
        } else {
            console.log('⚠️  WARNING: No spreadsheets found');
            console.log('   This means the service account has no spreadsheets shared with it');
        }
        console.log('============================================================\n');
        
    } catch (error) {
        console.log('\n============================================================');
        console.log('❌ TEST FAILED:', error.message);
        console.log('============================================================\n');
        process.exit(1);
    }
}

main();

File: test_login_flow.py
#!/usr/bin/env python3
"""
Test the complete staff login flow:
1. Login to Firebase with email/password
2. Get Firebase ID token
3. Call backend API to get service account
4. Use service account to get Google OAuth token
5. List spreadsheets using Google Drive API
"""

import requests
import json
import base64
import time
from datetime import datetime, timedelta
import jwt

# Configuration from .env.e2e
FIREBASE_API_KEY = "AIzaSyDsmavUPO_LoF6QxTz9x9HlPsdTWT8RdSs"
EMAIL = "chongjinheng@gmail.com"
PASSWORD = "123456"
BACKEND_API_URL = "https://wisata-dental.fly.dev"

def firebase_login(email, password):
    """Step 1: Login to Firebase and get ID token"""
    print("\n=== Step 1: Firebase Login ===")
    url = f"https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key={FIREBASE_API_KEY}"
    
    payload = {
        "email": email,
        "password": password,
        "returnSecureToken": True
    }
    
    response = requests.post(url, json=payload)
    print(f"Status: {response.status_code}")
    
    if response.status_code != 200:
        print(f"Error: {response.text}")
        return None
    
    data = response.json()
    id_token = data.get("idToken")
    print(f"✓ Firebase ID Token obtained (length: {len(id_token)})")
    return id_token

def get_service_account(firebase_token):
    """Step 2: Call backend API to get service account"""
    print("\n=== Step 2: Get Service Account from Backend ===")
    url = f"{BACKEND_API_URL}/api/auth/service-account"
    
    headers = {
        "Authorization": f"Bearer {firebase_token}",
        "Content-Type": "application/json"
    }
    
    response = requests.post(url, headers=headers)
    print(f"Status: {response.status_code}")
    
    if response.status_code != 200:
        print(f"Error: {response.text}")
        return None
    
    data = response.json()
    service_account = data.get("serviceAccount")
    print(f"✓ Service Account obtained: {service_account.get('client_email')}")
    return service_account

def get_google_access_token(service_account):
    """Step 3: Use service account to get Google OAuth token"""
    print("\n=== Step 3: Get Google OAuth Token ===")
    
    # Create JWT
    now = int(time.time())
    expiry = now + 3600
    
    jwt_payload = {
        "iss": service_account["client_email"],
        "sub": service_account["client_email"],
        "aud": "https://oauth2.googleapis.com/token",
        "iat": now,
        "exp": expiry,
        "scope": "https://www.googleapis.com/auth/spreadsheets https://www.googleapis.com/auth/drive.readonly"
    }
    
    # Sign JWT with private key
    private_key = service_account["private_key"]
    
    token = jwt.encode(jwt_payload, private_key, algorithm="RS256")
    
    # Exchange JWT for access token
    url = "https://oauth2.googleapis.com/token"
    payload = {
        "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
        "assertion": token
    }
    
    response = requests.post(url, data=payload)
    print(f"Status: {response.status_code}")
    
    if response.status_code != 200:
        print(f"Error: {response.text}")
        return None
    
    data = response.json()
    access_token = data.get("access_token")
    print(f"✓ Google Access Token obtained (length: {len(access_token)})")
    print(f"  Token type: {data.get('token_type')}")
    print(f"  Expires in: {data.get('expires_in')} seconds")
    return access_token

def list_spreadsheets(access_token):
    """Step 4: List spreadsheets using Google Drive API"""
    print("\n=== Step 4: List Spreadsheets ===")
    
    query = "mimeType='application/vnd.google-apps.spreadsheet' and trashed=false"
    url = f"https://www.googleapis.com/drive/v3/files?q={requests.utils.quote(query)}&fields=files(id,name)"
    
    headers = {
        "Authorization": f"Bearer {access_token}",
        "Content-Type": "application/json"
    }
    
    response = requests.get(url, headers=headers)
    print(f"Status: {response.status_code}")
    
    if response.status_code != 200:
        print(f"Error: {response.text}")
        return []
    
    data = response.json()
    files = data.get("files", [])
    
    print(f"\n✓ Found {len(files)} spreadsheets:")
    for idx, file in enumerate(files, 1):
        print(f"  {idx}. {file['name']} (ID: {file['id']})")
    
    return files

def main():
    print("=" * 60)
    print("Testing Complete Staff Login Flow")
    print("=" * 60)
    
    # Step 1: Firebase login
    firebase_token = firebase_login(EMAIL, PASSWORD)
    if not firebase_token:
        print("\n❌ Firebase login failed")
        return
    
    # Step 2: Get service account
    service_account = get_service_account(firebase_token)
    if not service_account:
        print("\n❌ Failed to get service account from backend")
        return
    
    # Step 3: Get Google access token
    google_token = get_google_access_token(service_account)
    if not google_token:
        print("\n❌ Failed to get Google access token")
        return
    
    # Step 4: List spreadsheets
    spreadsheets = list_spreadsheets(google_token)
    
    print("\n" + "=" * 60)
    if spreadsheets:
        print("✅ SUCCESS: Service account can access spreadsheets!")
        print(f"   Total spreadsheets found: {len(spreadsheets)}")
    else:
        print("⚠️  WARNING: No spreadsheets found")
        print("   This could mean:")
        print("   1. Service account has no spreadsheets shared with it")
        print("   2. There's an issue with permissions")
    print("=" * 60)

if __name__ == "__main__":
    main()
